### PR TITLE
Fix selectiontag on script parameter

### PR DESCRIPTION
**Pull request by vinjana**
- Allow for "selectionTag" in filename patterns and output file tags
- Correctly processes selectionTag for onScriptParameter filename pattern matching
- Added feature toggle `StrictOnScriptParameterSelectionTag` with default `false`; with `false` default selection tag values always match; with `true`, which will be the Roddy 4 default selection tag values in output file declarations only match default selection tags in filename patterns
- Improved parsing of `--cvalues` command-line argument solving issues with colons ':' in configuration variables
- Removed ModifiedVariablePassing feature toggle
- Fixed a new regression with not quoting bash/map variables in the job-parameter files
- Lots of minor improvements of (error) output and behaviour
- Updated BatchEuphoria dependency
- Added `testRunVariableWidth` application property to change the width of the printing of job-parameter names. Defaults to old values 25.
- Added escaping of not already quoted variables containing double quotes

### CODE REVIEW

File: RoddyCore/src/de/dkfz/roddy/config/converters/BashConverter.groovy [hunk 1]
```diff
@@ -317,27 +318,29 @@ class BashConverter extends ConfigurationConverter {
...
+                return new StringBuilder("${declareVar} ${cv.id}=${evaluatedValue}".toString())
             } else if (cv.type && cv.type.toLowerCase() == CVALUE_TYPE_PATH) {
                 tmp = "${cv.toFile(context)}".toString()
+            } else if (evaluatedValue.startsWith("-") || evaluatedValue.startsWith("*")) {
+                tmp = "\"${evaluatedValue}\"".toString()
+            } else if (!isQuoted(evaluatedValue) && evaluatedValue =~ /"/) {
+                tmp = Service.escape(evaluatedValue)
+                logger.always("Bash variable '${cv.id}' contains double quote. Will be escaped! Please check: [${tmp}]")
+            } else if (quoteSomeScalarConfigValues && !isQuoted(evaluatedValue) && evaluatedValue =~ /[\s\t\n;<>&!#'`|?{}\[\]()~\r\f$]/) {
+                // TODO User de.dkfz.roddy.tools.shell.bash.Service to escape characters. Note that quoting is the wrong thing if there are double quotes in the string! This should be implemented end to end with correct parsing of configuration values into arrays (account for escapes!) and without intermediate casting into Strings for bashArray variables.
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/config/converters/BashConverter.groovy]: User => Use

File: RoddyCore/src/de/dkfz/roddy/config/loader/ConfigurationFactory.groovy [hunk 2]
```diff
@@ -697,7 +698,7 @@ class ConfigurationFactory {
         Class<FileObject> _cls = loadPatternClass(pkg, filename.@class.text(), BaseFile).x
         String scriptName = filename.@onTool.text()
         String pattern = filename.@pattern.text()
-        String selectionTag = extractAttributeText(filename, "selectiontag", Constants.DEFAULT)
+        String selectionTag = extractSelectionTag(filename)
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/config/loader/ConfigurationFactory.groovy]: XSD!

File: RoddyCore/src/de/dkfz/roddy/config/loader/ProcessingToolReader.groovy [hunk 3]
```diff
@@ -234,60 +221,68 @@ class ProcessingToolReader {
...
+                tsp = new ToolStringParameter(pName, extractAttributeText(child, 'cValueID'))
             }
 
             return tsp
         } else {
-            addLoadErr("The type attribute of a parameter was invalid (${type}) for tool ${toolID}\n" + ConfigurationFactory.ERROR_PRINTOUT_XML_LINEPREFIX + RoddyConversionHelperMethods.toFormattedXML(child, "\n" + ConfigurationFactory.ERROR_PRINTOUT_XML_LINEPREFIX))
+            addLoadErr("The type attribute of a parameter was invalid (${type}) for tool ${toolID}\n" +
+                    ConfigurationFactory.ERROR_PRINTOUT_XML_LINEPREFIX +
+                    RoddyConversionHelperMethods.toFormattedXML(child, "\n" +
+                            ConfigurationFactory.ERROR_PRINTOUT_XML_LINEPREFIX))
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/config/loader/ProcessingToolReader.groovy]: You could place the last ) on the new line, which will help Idea to format the document in a better way.

File: RoddyCore/src/de/dkfz/roddy/config/loader/ProcessingToolReader.groovy [hunk 4]
```diff
@@ -234,60 +221,68 @@ class ProcessingToolReader {
...
             return parseTuple(child, toolID)
-        } else if (type == "filegroup") {
+        } else if (type == 'filegroup') {
             return parseFileGroup(child, toolID)
         } else if (type == CVALUE_TYPE_STRING) {
-            ToolStringParameter.ParameterSetbyOptions setby = Enum.valueOf(ToolStringParameter.ParameterSetbyOptions.class, extractAttributeText(child, "setby", ToolStringParameter.ParameterSetbyOptions.callingCode.name()))
+            ToolStringParameter.ParameterSetbyOptions setby =
+                    Enum.valueOf(ToolStringParameter.ParameterSetbyOptions.class,
+                            extractAttributeText(child, 'setby',
+                                    ToolStringParameter.ParameterSetbyOptions.callingCode.name()))
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/config/loader/ProcessingToolReader.groovy]: Also here (see below), consider placing the closing braces on new lines to help Idea with formatting.

File: RoddyCore/src/de/dkfz/roddy/knowledge/files/BaseFile.groovy [hunk 5]
```diff
@@ -853,10 +857,21 @@ abstract class BaseFile<FS extends FileStageSettings> extends FileObject {
         FilenamePattern appliedPattern = availablePatterns.find {
             FilenamePattern _fp ->
                 OnScriptParameterFilenamePattern fp = _fp as OnScriptParameterFilenamePattern
-                boolean parameterFound = fp.calledParameterID == helper.parameterID
+                boolean parameterFound = fp.parameterID == helper.parameterID
                 boolean scriptValid = fp.toolID ? fp.toolID == helper.toolID : true
-
-                return parameterFound && scriptValid
+                boolean selectionTagsMatch
+                if (Roddy.getFeatureToggleValue(FeatureToggles.StrictParameterSelectionTagEquality)) {
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/knowledge/files/BaseFile.groovy]: Strong and Weak match for other pattern types as well?

File: RoddyCore/src/de/dkfz/roddy/knowledge/files/BaseFile.groovy [hunk 6]
```diff
@@ -853,10 +857,21 @@ abstract class BaseFile<FS extends FileStageSettings> extends FileObject {
...
+                    // Strict matching: default only matches default
+                    selectionTagsMatch = fp.selectionTag == selectionTag
+                } else {
+                    // Weak matching: default always matches
+                    selectionTagsMatch =
```
Comment (dankwart-de) [inline RoddyCore/src/de/dkfz/roddy/knowledge/files/BaseFile.groovy]: Consider:
```Java
def fpSelectionTag = fp.selectionTag ?: FilenamePattern.DEFAULT_SELECTIONTAG
selectionTag = selectionTag?: FilenamePattern.DEFAULT_SELECTIONTAG
selectionTagsMatch = FilenamePattern.DEFAULT_SELECTIONTAG == fp.selectionTag ||
                                    FilenamePattern.DEFAULT_SELECTIONTAG == selectionTag ||
                                    fp.selectionTag == selectionTag
```

File: RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy [hunk 7]
```diff
@@ -72,33 +75,38 @@ class BaseFileTest {
         FileSystemAccessProvider.initializeProvider(true)
         //Setup plugins and default configuration folders
         LibrariesFactory.initializeFactory(true)
-        LibrariesFactory.getInstance().loadLibraries(LibrariesFactory.buildupPluginQueue(LibrariesFactoryTest.callLoadMapOfAvailablePlugins(), "DefaultPlugin").values() as List)
-        ConfigurationFactory.initialize(LibrariesFactory.getInstance().getLoadedPlugins().collect { it -> it.getConfigurationDirectory() })
+        LibrariesFactory.instance.loadLibraries(LibrariesFactory.
+                buildupPluginQueue(LibrariesFactoryTest.callLoadMapOfAvailablePlugins(), 'DefaultPlugin').values() as List)
```
Comment (dankwart-de) [inline RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy]: Maybe put the whole LibrariesFactory.buildupPluginQueue in a new line?
=>
```Java
 LibrariesFactory.instance.loadLibraries(
    LibrariesFactory.buildupPluginQueue(
        LibrariesFactoryTest.callLoadMapOfAvailablePlugins(), 'DefaultPlugin'
    ).values() as List
)
        
```

File: RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy [hunk 8]
```diff
@@ -72,33 +75,38 @@ class BaseFileTest {
         FileSystemAccessProvider.initializeProvider(true)
         //Setup plugins and default configuration folders
         LibrariesFactory.initializeFactory(true)
-        LibrariesFactory.getInstance().loadLibraries(LibrariesFactory.buildupPluginQueue(LibrariesFactoryTest.callLoadMapOfAvailablePlugins(), "DefaultPlugin").values() as List)
-        ConfigurationFactory.initialize(LibrariesFactory.getInstance().getLoadedPlugins().collect { it -> it.getConfigurationDirectory() })
+        LibrariesFactory.instance.loadLibraries(LibrariesFactory.
```
Comment (dankwart-de) [inline RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy]: Consider using LF, if you also use CF

File: RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy [hunk 9]
```diff
@@ -72,33 +75,38 @@ class BaseFileTest {
...
+                if (tool == 'wrapinScript')
                     return super.getSourceToolPath(tool)
-                return new File("/tmp/RoddyTests/RoddyTestScript_ExecutionServiceTest.sh")
+                return new File('/tmp/RoddyTests/RoddyTestScript_ExecutionServiceTest.sh')
             }
         }
-        syntheticTestFileClass = LibrariesFactory.getInstance().loadRealOrSyntheticClass("FileWithFileStage", BaseFile.class as Class<FileObject>)
-        ToolEntry toolEntry = new ToolEntry("RoddyTests", "RoddyTests", "RoddyTestScript_ExecutionServiceTest.sh")
-        toolEntry.getOutputParameters(mockupConfig).add(new ToolFileParameter(syntheticTestFileClass, null, "TEST", new ToolFileParameterCheckCondition(true)))
+        syntheticTestFileClass = LibrariesFactory.getInstance().loadRealOrSyntheticClass('FileWithFileStage', BaseFile.class as Class<FileObject>)
```
Comment (dankwart-de) [inline RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy]: Assign LibrariesFactory.instance to a field

File: docs/installationGuide.rst [hunk 10]
```diff
@@ -80,6 +80,10 @@ Roddy 2.3
 
 Please see `Versioning`_ for information about how to mix different versions of Roddy in the same directory.
 
+You will need an :doc:`config/applicationProperties` file with the configuration for accessing your computing environment.
+
+Note that the Roddy 2.4 releases are actually Roddy 3 pre-releases. Please use Roddy 3 instead.
+
 Roddy 3
```
Comment (dankwart-de) [inline docs/installationGuide.rst]: Consider a descending order for Roddy versions.

File: RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy (hunk 9 above)
Comment (vinjana) [reply RoddyCore/test/src/de/dkfz/roddy/knowledge/files/BaseFileTest.groovy]: Improvements are only marginal. Note that the factory initializers of CF and LF are different: CF does not return the singleton and LF's has a lengthly name. We could introduce ConfigurationFactory.initializeFactory (lengthly) to return a value.

What actually about importing the CFs static fields and static methods and safe even the chars for the CF. prefix?

**Pull request merged**